Bugfix/spack spec: read and use the environment concretizer:unification option

tldahlgren opened the pull request:
Fixed #38240 

This PR reads the `concretizer:unify` setting from the environment file (`spack.yaml`), if present, and uses the value it provides instead of the (default) configuration (e.g., https://spack.readthedocs.io/en/latest/build_settings.html#concretizer-options) so that `spack spec` follows the proper concretization path.

tldahlgren committed:
Bugfix: spack.yaml concretizer:unify needs to be read and used

tldahlgren committed:
Optional: add environment test to ensure configuration scheme is used

alalazo commented on lib/spack/spack/environment/environment.py:
```diff
@@ -859,7 +859,13 @@ def _construct_state_from_manifest(self):
 
         # Retrieve unification scheme for the concretizer
-        self.unify = spack.config.get("concretizer:unify", False)
+        concretizer = configuration.get("concretizer", {})
+        unify = None
+        for entry, value in concretizer.items():
+            if entry == "unify":
+                unify = value
+                break
+        self.unify = unify if unify is not None else spack.config.get("concretizer:unify", False)
```
I'd like to double check this one. My hypothesis is that this function is called with an implicit assumption on the environment being active, in which case `spack.config` has the environment config scope added at some point of the stack.

If the issues comes from scripting instead, and we don't want to make this assumption, checking the `spack.yaml` file first, and then the configuration, might be wrong - since from a script we can still add an arbitrary amount of config layers on top of the spack.yaml single file scope

tldahlgren replied:
Okay.  The problem we had was that the value in `spack.yaml` did not appear to be read or used at all for the active environment, or at least not used when it counted.

alalazo replied:
Based on the discussion in #38240, I'd rather solve this with:
```diff
diff --git a/lib/spack/spack/environment/environment.py b/lib/spack/spack/environment/environment.py
index 82810e3b7d..6b8006ff2e 100644
--- a/lib/spack/spack/environment/environment.py
+++ b/lib/spack/spack/environment/environment.py
@@ -758,7 +758,8 @@ def __init__(self, manifest_dir: Union[str, pathlib.Path]) -> None:
 
         self.txlock = lk.Lock(self._transaction_lock_path)
 
-        self.unify = None
+        self._unify = None
+
         self.new_specs: List[Spec] = []
         self.new_installs: List[Spec] = []
         self.views: Dict[str, ViewDescriptor] = {}
@@ -782,6 +783,16 @@ def __init__(self, manifest_dir: Union[str, pathlib.Path]) -> None:
             self.manifest = EnvironmentManifestFile(manifest_dir)
             self._read()
 
+    @property
+    def unify(self):
+        if self._unify is None:
+            self._unify = spack.config.get("concretizer:unify", False)
+        return self._unify
+
+    @unify.setter
+    def unify(self, value):
+        self._unify = value
+
     def __reduce__(self):
         return _create_environment, (self.path,)
 
@@ -843,9 +854,6 @@ def _construct_state_from_manifest(self):
         else:
             self.views = {}
 
-        # Retrieve unification scheme for the concretizer
-        self.unify = spack.config.get("concretizer:unify", False)
-
         # Retrieve dev-build packages:
         self.dev_specs = copy.deepcopy(env_configuration.get("develop", {}))
         for name, entry in self.dev_specs.items():

```
The rationale is that the issue is due to Spack reading `self.unify` too early - so we can fix that by delaying when we read it. A more long term fix would be to _inject_ the global configuration into the environment and avoid relying on global variables - but that is a major refactoring of the `Environment` class which is not required for a bugfix imho.

I wouldn't go with this solution since:
1. There are other methods in the `Environment` class that expect the global configuration to contain the manifest `SingleFileScope`, for instance `_concretize_separately`. I don't see a reason to special case this method.
2. We shouldn't be managing configuration scopes manually, when we have the `Configuration` class for that purpose

alalazo replied:
@tldahlgren Any thoughts on :point_up: If you want, I can push a commit here in that sense.

tldahlgren replied:
@alalazo Yes please do push a commit so you are credited for the work.

alalazo committed:
Activate environment in unit tests

A more proper solution would be to keep
an environment instance configuration as
an attribute, but that is a bigger refactor

alalazo committed:
Delay evaluation of Environment.unify

alalazo committed:
Slightly simplify unit tests

alalazo commented:
@tldahlgren Let me know if you agree with the latest changes

tldahlgren commented:
> @tldahlgren Let me know if you agree with the latest changes

Looks like it doesn't resolve the original problem described in #38240.

```
$ spack config --scope user add concretizer:unify:false
$ spack env activate --temp
$ spack add zlib@1.2.13
==> Adding zlib@1.2.13 to environment /var/tmp/dahlgren/spack-w23okd3k
$ spack add zlib@1.2.12
==> Adding zlib@1.2.12 to environment /var/tmp/dahlgren/spack-w23okd3k
$ spack config add concretizer:unify:true 
$ spack concretize
==> Error: concretization failed for the following reasons:

   1. Cannot select a single "version" for package "zlib"
   2. Cannot satisfy 'zlib@1.2.13'
   3. Cannot satisfy 'zlib@1.2.12'. You could consider setting `concretizer:unify` to `when_possible` or `false` to allow multiple versions of some packages.
$ spack spec
==> Error: concretization failed for the following reasons:

   1. Cannot select a single "version" for package "zlib"
   2. Cannot satisfy 'zlib@1.2.13'
   3. Cannot satisfy 'zlib@1.2.12'. You could consider setting `concretizer:unify` to `when_possible` or `false` to allow multiple versions of some packages.
```
